File: asciinema/asciicast.py
import sys
import json
import json.decoder
import urllib.request
import urllib.error
import html.parser
from .stdout import Stdout


class Asciicast:

    def __init__(self, stdout, width, height, duration, command=None, title=None, term=None, shell=None):
        self.stdout = stdout
        self.width = width
        self.height = height
        self.duration = duration
        self.command = command
        self.title = title
        self.term = term
        self.shell = shell

    def save(self, path):
        stdout = list(map(lambda frame: [round(frame[0], 6), frame[1]], self.stdout.frames))
        duration = round(self.duration, 6)
        attrs = {
            "version": 1,
            "width": self.width,
            "height": self.height,
            "duration": duration,
            "command": self.command,
            "title": self.title,
            "env": {
                "TERM": self.term,
                "SHELL": self.shell
            },
            "stdout": stdout
        }

        with open(path, "w") as f:
            f.write(json.dumps(attrs, ensure_ascii=False, indent=2))


# asciinema play file.json
# asciinema play https://asciinema.org/a/123.json
# asciinema play https://asciinema.org/a/123
# asciinema play ipfs://ipfs/QmbdpNCwqeZgnmAWBCQcs8u6Ts6P2ku97tfKAycE1XY88p
# asciinema play -


class LoadError(Exception):
    pass


class Parser(html.parser.HTMLParser):
    def __init__(self):
        html.parser.HTMLParser.__init__(self)
        self.url = None

    def handle_starttag(self, tag, attrs_list):
        # look for <link rel="alternate" type="application/asciicast+json" href="https://...json">
        if tag == 'link':
            attrs = {}
            for k, v in attrs_list:
                attrs[k] = v

            if attrs.get('rel') == 'alternate' and attrs.get('type') == 'application/asciicast+json':
                self.url = attrs.get('href')


def fetch(url):
    if url.startswith("ipfs:/"):
        url = "https://ipfs.io/%s" % url[6:]
    elif url.startswith("fs:/"):
        url = "https://ipfs.io/%s" % url[4:]

    if url == "-":
        return sys.stdin.read()

    if url.startswith("http:") or url.startswith("https:"):
        response = urllib.request.urlopen(url)
        data = response.read().decode(errors='replace')

        content_type = response.headers['Content-Type']
        if content_type and content_type.startswith('text/html'):
            parser = Parser()
            parser.feed(data)
            url = parser.url

            if not url:
                raise LoadError("""<link rel="alternate" type="application/asciicast+json" href="..."> not found in fetched HTML document""")

            return fetch(url)

        return data

    with open(url, 'r') as f:
        return f.read()


def load(filename):
    try:
        attrs = json.loads(fetch(filename))

        if type(attrs) != dict:
            raise LoadError('unsupported asciicast format')

        return Asciicast(
            attrs['stdout'],
            attrs['width'],
            attrs['height'],
            attrs['duration'],
            attrs['command'],
            attrs['title']
        )
    except (OSError, urllib.error.HTTPError) as e:
        raise LoadError(str(e))
    except json.decoder.JSONDecodeError as e:
        raise LoadError('JSON decoding error: ' + str(e))
    except KeyError as e:
        raise LoadError('asciicast is missing key ' + str(e))

File: asciicast/asciicast.go
package asciicast

import (
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"strings"

	"github.com/asciinema/asciinema/Godeps/_workspace/src/golang.org/x/net/html"
)

type Env struct {
	Term  string `json:"TERM"`
	Shell string `json:"SHELL"`
}

type Duration float64

func (d Duration) MarshalJSON() ([]byte, error) {
	return []byte(fmt.Sprintf(`%.6f`, d)), nil
}

type Asciicast struct {
	Version  int      `json:"version"`
	Width    int      `json:"width"`
	Height   int      `json:"height"`
	Duration Duration `json:"duration"`
	Command  string   `json:"command"`
	Title    string   `json:"title"`
	Env      *Env     `json:"env"`
	Stdout   []Frame  `json:"stdout"`
}

func NewAsciicast(width, height int, duration float64, command, title string, frames []Frame, env map[string]string) *Asciicast {
	return &Asciicast{
		Version:  1,
		Width:    width,
		Height:   height,
		Duration: Duration(duration),
		Command:  command,
		Title:    title,
		Env:      &Env{Term: env["TERM"], Shell: env["SHELL"]},
		Stdout:   frames,
	}
}

func Save(asciicast *Asciicast, path string) error {
	bytes, err := json.MarshalIndent(asciicast, "", "  ")
	if err != nil {
		return err
	}

	err = ioutil.WriteFile(path, bytes, 0644)
	if err != nil {
		return err
	}

	return nil
}

// asciinema play file.json
// asciinema play https://asciinema.org/a/123.json
// asciinema play https://asciinema.org/a/123
// asciinema play ipfs://ipfs/QmbdpNCwqeZgnmAWBCQcs8u6Ts6P2ku97tfKAycE1XY88p
// asciinema play -

func getAttr(t *html.Token, name string) string {
	for _, a := range t.Attr {
		if a.Key == name {
			return a.Val
		}
	}

	return ""
}

func extractJSONURL(htmlDoc io.Reader) (string, error) {
	z := html.NewTokenizer(htmlDoc)

	for {
		tt := z.Next()

		switch {
		case tt == html.ErrorToken:
			return "", fmt.Errorf("expected alternate <link> not found in fetched HTML document")
		case tt == html.StartTagToken:
			t := z.Token()

			if t.Data == "link" && getAttr(&t, "rel") == "alternate" && getAttr(&t, "type") == "application/asciicast+json" {
				return getAttr(&t, "href"), nil
			}
		}
	}
}

func getSource(url string) (io.ReadCloser, error) {
	var source io.ReadCloser
	var isHTML bool
	var err error

	if strings.HasPrefix(url, "ipfs:/") {
		url = fmt.Sprintf("https://ipfs.io/%v", url[6:])
	} else if strings.HasPrefix(url, "fs:/") {
		url = fmt.Sprintf("https://ipfs.io/%v", url[4:])
	}

	if url == "-" {
		source = os.Stdin
	} else if strings.HasPrefix(url, "http://") || strings.HasPrefix(url, "https://") {
		resp, err := http.Get(url)

		if err != nil {
			return nil, err
		}

		if resp.StatusCode != 200 {
			resp.Body.Close()
			return nil, fmt.Errorf("got status %v when requesting %v", resp.StatusCode, url)
		}

		source = resp.Body

		if strings.HasPrefix(resp.Header.Get("Content-Type"), "text/html") {
			isHTML = true
		}
	} else {
		source, err = os.Open(url)
		if err != nil {
			return nil, err
		}

		if strings.HasSuffix(url, ".html") {
			isHTML = true
		}
	}

	if isHTML {
		defer source.Close()
		url, err = extractJSONURL(source)
		if err != nil {
			return nil, err
		}

		return getSource(url)
	}

	return source, nil
}

func Load(url string) (*Asciicast, error) {
	source, err := getSource(url)
	if err != nil {
		return nil, err
	}
	defer source.Close()

	dec := json.NewDecoder(source)
	asciicast := &Asciicast{}

	if err = dec.Decode(asciicast); err != nil {
		return nil, err
	}

	return asciicast, nil
}
